### PR TITLE
Language Model Redesign

### PR DESCRIPTION
**Pull request by dcgaines**
# Overview

This PR reworked the existing GPT-2 language model implementation to fix several critical bugs. The new CasualLanguageModel class fixes these bugs and allows for the insertion of any causal model from HuggingFace (or a locally trained one). This PR also includes the KenLMLanguageModel class that implements an n-gram model and the MixtureLanguageModel class that allows for the mixture of two or more other models. 

## Ticket

https://www.pivotaltracker.com/story/show/183975978
https://www.pivotaltracker.com/story/show/184017969
https://www.pivotaltracker.com/story/show/184589162
https://www.pivotaltracker.com/story/show/184365440

## Contributions

- lm_eval.py script for the evaluation of perplexity and prediction time for any given language model
- mixture_tuning.py script for the optimization of mixture weights for use in the MixtureLanguageModel
- CausalLanguageModel class, a generic language model class that can be used with many different causal models from HuggingFace
- KenLMLanguageModel class, an n-gram language model class
- MixtureLanguageModel class, which allows for the mixture of the outputs of two or more other models
- Creation of lm_params.json to accommodate the usage of the new language model classes. This file is copied into the session data directory alongside parameters.json
- Deprecated old GPT-2 implementation due to critical bugs. It is replaced by the CausalLanguageModel class, which can use gpt2 from HuggingFace

## Test

- Wrote test suites for each language model class based on the pre-existing tests for the previous GPT2LanguageModel class. Expanded tests where necessary to ensure proper performance. 
- Tested locally using a fake LSL server
- Tested on new home use machine by OHSU development and clinical teams.

## Documentation

- Updated the language module README as well as a few slight adjustments to the overall README. 

## Changelog

- Is the CHANGELOG.md updated with your detailed changes? Not yet

### CODE REVIEW

File: .gitignore
Comment (tab-cmd) [inline .gitignore:31]: Could this be updated to *.arpa? Or do we want those file types uploadable?

File: bcipy/display/paradigm/vep/display.py
Comment (tab-cmd) [inline bcipy/display/paradigm/vep/display.py:13]: We've gone back and forth over the years about whether this should live in language or not (as well as the alphabet; better-named symbol set likely). The rationale for leaving it in helpers is to encapsulate the dependencies between modules as best we can. As the language module becomes more feature-rich, we may need to consider other patterns.

File: bcipy/helpers/exceptions.py
Comment (tab-cmd) [inline bcipy/helpers/exceptions.py]: How about `InvalidLangaugeModelException`? I was going to advocate for a generic, which the signal module could then use, but it has its own handling so for now we can afford to be explicit.

File: bcipy/helpers/parameters.py
Comment (tab-cmd) [inline bcipy/helpers/parameters.py]: This bit of code is to handle parameter lists... we should break this out into a function since there could be many different types of lists if we start supporting them... @lawhead may know a better way to deal with this using the base class. 

If there's not a straightforward way to represent, and these checks are needed, we could pull out some common elements like:
```
if entry['value'] == "":
    return None

list_items = entry['value'].split(", ")

# check the type and cast as before
if cast == List[float]:
    ...

...
```

Also, do we want to support this case or fail as invalid parameters?

```
if entry['value'] == "":
    return None
```

File: bcipy/language/README.md
Comment (tab-cmd) [inline bcipy/language/README.md:46]: Nice documentation! We could use some description of the new structure. Ex. lms, sets,

Comment (tab-cmd) [reply bcipy/language/README.md:46]: We can add another ticket in the backlog for this, but we could use better documentation on integrating a new model. You had to figure a great deal out by yourself, but a walkthrough would be more accessible.

File: bcipy/language/demo/autocase_gpt2.py
Comment (tab-cmd) [inline bcipy/language/demo/autocase_gpt2.py]: rename, `demo_autocase_gpt2.py`

File: bcipy/language/demo/demo_kenlm.py
Comment (tab-cmd) [inline bcipy/language/demo/demo_kenlm.py:90]: These demos look nice!

File: bcipy/language/lms/zebras.txt
Comment (tab-cmd) [inline bcipy/language/lms/zebras.txt]: Do you think this file is necessary? We could generate it on the fly or code it as a string since it is only one line.

File: bcipy/language/model/causal.py
Comment (tab-cmd) [inline bcipy/language/model/causal.py:27]: We currently only use cpu. Do we want to support the others right now? Adding them incrementally as needed may be better to avoid unknown errors.

File: bcipy/parameters/lm_params.json
Comment (tab-cmd) [inline bcipy/parameters/lm_params.json]: These could live closer together in the mixture params since they are directly related, especially if we already support a list of dictionaries. 

```
{
"mixture": 
 "recommended_values": [
      "{}, {}",
      "{'lang_model_name': 'gpt2', 'weight': 0.1, 'type': 'CAUSAL'}, "{ 'weight': 0.9, 'type': 'KENLM'},"
}
```

File: requirements.txt
Comment (tab-cmd) [inline requirements.txt]: Does this work with PyPi install now?

File: scripts/python/lm_eval.py
Comment (tab-cmd) [inline scripts/python/lm_eval.py:22]: This is a great script to have. I'm opinionated about scripts living here in the codebase, though. Often, it's where untested (as in unittests and linting!) and feature-rich code gets rooted and stays. It gets updated less reliably and goes stale even if it becomes a critical part of our workflow. I have been bitten several times and hope desperately to deprecate this in the repo. 

I'd be happy to keep it in here for now if we make another ticket to pull out what we can into the repo as supported code (and then use a demo to evoke it!)

File: scripts/python/mixture_tuning.py
Comment (tab-cmd) [inline scripts/python/mixture_tuning.py]: rm?

File: scripts/python/mixture_tuning.py
Comment (tab-cmd) [inline scripts/python/mixture_tuning.py]: rm

File: scripts/python/mixture_tuning.py
Comment (tab-cmd) [inline scripts/python/mixture_tuning.py]: Not req

**Review by tab-cmd — CHANGES_REQUESTED** [review]
Here is my initial review! Significant work here. Let's wait for @lawhead to review the models and patterns before merging or addressing noncosmetic items!

File: requirements.txt
Comment (tab-cmd) [reply requirements.txt]: I get the following on Windows:

```
python/kenlm.cpp(710): fatal error C1083: Cannot open include file: 'python/score_sentence.hh': No such file or directory
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.33.31629\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2
      [end of output
```

We will need to fix this before merging or remove it and include manual install instructions -- with this approach, we could add a dep import check and a nice message to the user when they try using it without proper installation.

File: .gitignore
Comment (dcgaines) [reply .gitignore:31]: There's a few different models of different sizes. I think we have the tiny model currently in the repo, while the large model is ignored and can be downloaded from our website.

File: bcipy/display/paradigm/vep/display.py
Comment (dcgaines) [reply bcipy/display/paradigm/vep/display.py:13]: We needed to move those things so that our evaluation scripts didn't need to import psychopy, which was preventing them from running headless with pygame. I believe psychopy was being imported by whatever contained alphabet(), SPACE_CHAR, and BACKSPACE_CHAR previously. Is the language module not the primary module that uses those constants and alphabet()?

Comment (lawhead) [reply bcipy/display/paradigm/vep/display.py:13]: I agree that it's an issue needing to import psychopy to use these. However, the alphabet symbols are a cross-cutting concern used throughout the code base including displays, the decision maker, tasks, language model, etc. Also the alphabet is not just comprised of characters but may be images or other icons (we previously did an icon-based experiment). Maybe they should go in their own helper (bcipy/helpers/symbols.py)?

File: bcipy/helpers/parameters.py
Comment (lawhead) [inline bcipy/helpers/parameters.py]: The parameters module was structured this way to support configuration parameters that could be modified by clinicians using a GUI interface. It was intentionally limited to simple types. If you need more structured data, I would recommend putting that in a separate json file and loading it using the python `json` module. See the `devices.json` file and the `acquisition/devices.py` module as an example of this approach. I would prefer not to have any changes to this module.

File: bcipy/parameters/lm_params.json
Comment (lawhead) [inline bcipy/parameters/lm_params.json:1]: As I mentioned in the parameters module, this configuration doesn't necessarily need to be formatted as a Parameters file because it's not edited by the clinicians through a GUI. So feel free to choose a format that is specific to your needs. Let me know if you want to work on this together or if it would be useful to have specific suggestions.

File: bcipy/language/model/mixture.py
Comment (lawhead) [inline bcipy/language/model/mixture.py]: You could use a list comprehension here:

    pred_list = [dict(model.predict(evidence)) for model in models]

File: bcipy/language/model/causal.py
Comment (lawhead) [inline bcipy/language/model/causal.py]: https://docs.python.org/3/library/stdtypes.html#str.replace

    s.replace(' ', SPACE_CHAR)

I would probably just remove this method, since it's redundant.

File: bcipy/language/model/causal.py
Comment (lawhead) [inline bcipy/language/model/causal.py]: We try to delete unused code to keep things tidy. It's always in version control if we need it back.

**Review by lawhead — CHANGES_REQUESTED** [review]
This is a lot of work! Thanks for your effort on this. The main things I would like to revisit would be parameters and maybe the location of symbols. See my comments below.

File: bcipy/language/model/causal.py
Comment (dcgaines) [reply bcipy/language/model/causal.py:27]: Keith and I use the other with the lm_eval script.

File: bcipy/helpers/exceptions.py
Comment (dcgaines) [reply bcipy/helpers/exceptions.py]: Renamed!

File: requirements.txt
Comment (dcgaines) [reply requirements.txt]: Added try/except surrounding import with installation instructions

File: bcipy/display/paradigm/vep/display.py
Comment (dcgaines) [reply bcipy/display/paradigm/vep/display.py:13]: Created a specific helper for these imports

File: bcipy/language/demo/autocase_gpt2.py
Comment (dcgaines) [reply bcipy/language/demo/autocase_gpt2.py]: This script was meant to be removed entirely

File: scripts/python/lm_eval.py
Comment (dcgaines) [reply scripts/python/lm_eval.py:22]: Ticket created! 
https://www.pivotaltracker.com/story/show/184794336

File: bcipy/language/README.md
Comment (dcgaines) [reply bcipy/language/README.md:46]: Added documentation for the module structure. I'm not quite sure what you mean be integrating a new model, could you elaborate a little more?

File: bcipy/helpers/parameters.py
Comment (dcgaines) [reply bcipy/helpers/parameters.py]: Reverted changes to this module.

File: bcipy/helpers/parameters.py
Comment (dcgaines) [reply bcipy/helpers/parameters.py]: Fixed!

File: .github/workflows/main.yml
Comment (tab-cmd) [inline .github/workflows/main.yml]: This line is probably not needed for the builds

File: bcipy/helpers/tests/resources/mock_session/parameters.json
Comment (tab-cmd) [inline bcipy/helpers/tests/resources/mock_session/parameters.json:648]: It might be a better default to use UNIFORM? I could be swayed either way. Whatever you choose, update the helpTip with the correct `Default:`

File: bcipy/language/model/kenlm.py
Comment (tab-cmd) [inline bcipy/language/model/kenlm.py]: cleanup

File: bcipy/language/tests/test_language.py
Comment (tab-cmd) [inline bcipy/language/tests/test_language.py]: This should move to `helpers/tests/test_symbols.py`

File: .github/workflows/main.yml
Comment (tab-cmd) [inline .github/workflows/main.yml]: Is there a version we could/should pin?

File: bcipy/language/model/kenlm.py
Comment (tab-cmd) [inline bcipy/language/model/kenlm.py]: cleanup

File: bcipy/parameters/lm_params.json
Comment (tab-cmd) [inline bcipy/parameters/lm_params.json:5]: `*.arpa`?

**Review by tab-cmd — APPROVED** [review]
Some small cleanups are needed, and there is a typo in lm params? Otherwise, this is ready to merge and iterate on! @lawhead  will want to take another look

File: .github/workflows/main.yml
Comment (tab-cmd) [reply .github/workflows/main.yml]: We should add this to the installation instructions in the README.md!

Comment (dcgaines) [reply .github/workflows/main.yml]: I had it in the language module README, but I've added it to the general README as well.

File: bcipy/helpers/tests/resources/mock_session/parameters.json
Comment (dcgaines) [reply bcipy/helpers/tests/resources/mock_session/parameters.json:648]: I feel the default should be our best LM, which at this point is the mixture. I've updated the helpTip defaults accordingly

File: bcipy/parameters/lm_params.json
Comment (dcgaines) [reply bcipy/parameters/lm_params.json:5]: We've shifted to using the binary .kenlm files instead of the .arpa files since they have faster model load times. I believe I already updated all the documentation to point to the proper .kenlm file location

**Pull request merged**
